buildbot_nix/__init__.py

```diff
@@ -49,95 +65,439 @@ class BuildbotNixError(Exception):
-            props.setProperty("system", system, source)
-            props.setProperty("drv_path", drv_path, source)
-            props.setProperty("out_path", out_path, source)
-            props.setProperty("default_branch", self.project.default_branch, source)
-            # we use this to identify builds when running a retry
-            props.setProperty("build_uuid", str(uuid.uuid4()), source)
-
-            if error is not None:
-                props.setProperty("error", error, source)
-                triggered_schedulers.append((self.skipped_builds_scheduler, props))
-                continue
-
-            if job.get("isCached"):
-                triggered_schedulers.append((self.skipped_builds_scheduler, props))
-                continue
-
-            build_props.setProperty(f"{attr}-out_path", out_path, source)
-            build_props.setProperty(f"{attr}-drv_path", drv_path, source)
-
-            triggered_schedulers.append((self.builds_scheduler, props))
-        return triggered_schedulers
+        ss_for_trigger = self.prepare_sourcestamp_list_for_trigger()
+        scheduler_log: Log = yield self.addLog("scheduler")
+
+        # inject failed buildsteps for any failed eval jobs we got
+        overall_result = SUCCESS if not self.failed_jobs else util.FAILURE
+        if self.failed_jobs:
+            scheduler_log.addStdout("The following jobs failed to evaluate:\n")
+        for failed_job in self.failed_jobs:
+            scheduler_log.addStdout(f"\t- {failed_job.attr} failed eval\n")
+            brids, _ = yield self.schedule(ss_for_trigger, *self.schedule_eval_failure(failed_job, self.report_status))
+            self.brids.extend(brids)
+
+        # get all input derivations for every job as a dictionary
+        derivations_inputs: dict[str, set[str]] = {
+            derivation: set(info.inputDrvs.keys())
+            for derivation, info in self.all_derivations.items()
+        }
+
+        # get all job derivations
+        job_set = {job.drvPath for job in self.successful_jobs}
+
+        # restrict the set of input derivations for all jobs to those which themselves are jobs
+        job_closures = {
+            k: self.get_derivation_closure(k, derivations_inputs).intersection(job_set)
+            for k in job_set
+        }
+
+        # sort them according to their dependencies
+        build_schedule_order = self.sort_jobs_by_closures(
+            self.successful_jobs, job_closures
+        )
+
+        done: list[BuildTrigger.DoneJob] = []
+        scheduled: list[BuildTrigger.ScheduledJob] = []
+        while build_schedule_order or scheduled:
+            scheduler_log.addStdout("Scheduling...\n")
+
+            # check which jobs should be scheduled now
+            schedule_now = []
+            for build in list(build_schedule_order):
+                if job_closures.get(build.drvPath):
+                    pass
+                elif (
+                    failed_builds.check_build(build.drvPath)
```

Comment:
Could we also store timestamp in there for the future? Than we can garbage collect at some point.

Reply:
Maybe use something like json as the value i.e. `{"ts": 1233e131313131}` than we can also add fields.

Reply:
Could be done yes, will ad before merging